crates/noirc_evaluator/src/ssa/optim.rs

```diff
@@ -258,23 +242,23 @@ pub fn find_similar_mem_instruction(
     ins_id: NodeId,
     lhs: NodeId,
     rhs: NodeId,
-    anchor: &HashMap<node::Operation, VecDeque<NodeId>>,
+    prev_ins: &VecDeque<NodeId>,
 ) -> (NodeId, CseAction) {
     match op {
         node::Operation::Load(_) => {
-            for iter in anchor[&op].iter().rev() {
+            for iter in prev_ins.iter() {
                 if let Some(ins_iter) = ctx.try_get_instruction(*iter) {
                     match ins_iter.operator {
                         node::Operation::Load(_) => {
-                            if ins_iter.lhs == lhs {
+                            if !ctx.maybe_distinct(ins_iter.lhs, lhs) {
                                 return (*iter, CseAction::Replace);
                             }
                         }
                         node::Operation::Store(_) => {
-                            if ins_iter.rhs == lhs {
+                            if !ctx.maybe_distinct(ins_iter.rhs, lhs) {
                                 return (ins_iter.lhs, CseAction::Replace);
-                            } else {
-                                //TODO: If we know that ins.lhs value cannot be equal to ins_iter.rhs, we could continue instead
+                            }
+                            if ctx.maybe_equal(ins_iter.rhs, lhs) {
```

Comment:
The maybe_equal doesn't seem needed since the default action is to Keep anyway

Reply:
I think it is needed, if it is not maybe_equal, we continue the loop and we may get another action.